### PR TITLE
code refactor

### PR DESCRIPTION
**Pull request by parthshah51999**

### CODE REVIEW

File: src/toDoList.js [hunk 1]
```diff
@@ -2,39 +2,66 @@ import React from 'react';
...
       this.addItem = this.addItem.bind(this);
       this.displayList = this.displayList.bind(this);
+      this.removeItem = this.removeItem.bind(this);
   }
 
   addItem() {
-    let inputText = this.state.value;
-    this.state.items = [...this.state.items, inputText];
-    this.state.value = '';
+    let inputText = this.state.value; //object destructring
```
Comment (visha12l) [inline src/toDoList.js]: use only one let per scope

File: src/toDoList.js [hunk 2]
```diff
@@ -2,39 +2,66 @@ import React from 'react';
...
-    return (this.state.items.slice(0).reverse().map((item,i) => <li key={i}>{item}</li>));
+  removeItem(id) {
+    let newItems = this.state.items.filter((item, i) => i !== id);
+    this.setState({
+      items : newItems
+    });
+  }
+
+  displayList() {
+    return (<ul>{this.state.items.map((item, i) => {
```
Comment (visha12l) [inline src/toDoList.js]: code after return should be on new line

File: src/toDoList.js [hunk 3]
```diff
@@ -2,39 +2,66 @@ import React from 'react';
...
+      return (
+        <li key={i}>{item}
+        <button
+          onClick={this.removeItem.bind(this, i)}
+          type="button"
+          className="btn btn-primary">Remove
+        </button>
+      </li>);
+    })}</ul>);
```
Comment (visha12l) [inline src/toDoList.js]: move </ul> tag to new line

File: src/toDoList.js [hunk 4]
```diff
@@ -2,39 +2,66 @@ import React from 'react';
...
-    let items = this.state.items;
+  render() { //render is a react component method.
     return (
       <div>
         <div id="addTask">
-          <input className="input" size="40" type="text" placeholder="Enter Task" onChange={e => this.setState({ value: e.target.value })} value={this.state.value}/>
-          <button onClick={this.addItem} type="button" className="btn btn-primary">Add</button>
+          <input
+            className="input"
+            size="40"
```
Comment (visha12l) [inline src/toDoList.js]: remove size attribute

File: src/toDoList.js [hunk 5]
```diff
@@ -2,39 +2,66 @@ import React from 'react';
...
+      this.removeItem = this.removeItem.bind(this);
   }
 
   addItem() {
-    let inputText = this.state.value;
-    this.state.items = [...this.state.items, inputText];
-    this.state.value = '';
+    let inputText = this.state.value; //object destructring
+    let newItems = [...this.state.items, inputText];
+    //this.state.value = '';  dont change state variables directly... use setState instead
```
Comment (visha12l) [inline src/toDoList.js]: https://stackoverflow.com/questions/37755997/why-cant-i-directly-modify-a-components-state-really/37760774

File: src/toDoList.js [hunk 6]
```diff
@@ -38,10 +50,27 @@ class ToDoList extends React.Component {
...
+          onClick={this.editItem.bind(this, i)}
+          type="button"
+          className="btn btn-primary">Edit
+        </button>
       </li>);
     })}</ul>);
   }
 
+  doRenderAddOrUpdateButton() {
+    let buttonText = "Add";
```
Comment (visha12l) [inline src/toDoList.js]: use single quotes

File: src/toDoList.js [hunk 7]
```diff
@@ -38,10 +50,27 @@ class ToDoList extends React.Component {
...
+    if(this.state.isUpdate){buttonText = "Edit";}
```
Comment (visha12l) [inline src/toDoList.js]: space required after if

File: src/toDoList.js [hunk 8]
```diff
@@ -5,20 +5,32 @@ class ToDoList extends React.Component {
...
-    let newItems = [...this.state.items, inputText];
+    let newItems = this.state.items;
+    {this.state.isUpdate ? newItems[this.state.id] = inputText : newItems = [...this.state.items, inputText]};
     //this.state.value = '';  dont change state variables directly... use setState instead
     this.setState({
       value : '',
-      items : newItems
+      items : newItems,
+      isUpdate : false,
+      id : 9999
```
Comment (visha12l) [inline src/toDoList.js]: why you are using static id ??

File: src/toDoList.js [hunk 9]
```diff
@@ -5,20 +5,32 @@ class ToDoList extends React.Component {
...
       };
       this.addItem = this.addItem.bind(this);
       this.displayList = this.displayList.bind(this);
       this.removeItem = this.removeItem.bind(this);
   }
 
+  editItem(id) {
+    this.setState({
+      isUpdate : true,
+      id : id
```
Comment (visha12l) [inline src/toDoList.js]: use enhanced object literal

File: src/toDoList.js [hunk 10]
```diff
@@ -5,20 +5,32 @@ class ToDoList extends React.Component {
...
+        value: '',
+        isUpdate: false,
+        id: 9999
       };
       this.addItem = this.addItem.bind(this);
       this.displayList = this.displayList.bind(this);
       this.removeItem = this.removeItem.bind(this);
   }
 
+  editItem(id) {
```
Comment (visha12l) [inline src/toDoList.js]: space required before opening bracket

File: src/toDoList.js [hunk 11]
```diff
@@ -2,39 +2,153 @@ import React from 'react';
...
+        onClick={this.addItem}
+        type="button"
+        className="btn btn-primary">{buttonText}
+      </button>
+    );
   }
 
-  render() {
-    let items = this.state.items;
+  doRenderSearchBar() {
```
Comment (abp437) [inline src/toDoList.js]: Why is this moved to a separate function? No extra logic is involved here.

File: src/toDoList.js [hunk 12]
```diff
@@ -2,39 +2,153 @@ import React from 'react';
...
+    />
+    );
+  }
+
+  onChangeFunctions(action) {
+    return (
+      action === 'search' ?
+      e => {
+        let searchResults = null;
+        searchResults = this.state.items.filter((value, i) => value.includes(e.target.value));
```
Comment (abp437) [inline src/toDoList.js]: Get rid of includes. It's not supported in IE. Even when coding in JS, it should be cross browser.

File: src/toDoList.js [hunk 13]
```diff
@@ -2,39 +2,153 @@ import React from 'react';
...
+      value={this.state.searchValue}
+    />
+    );
+  }
+
+  onChangeFunctions(action) {
+    return (
+      action === 'search' ?
+      e => {
+        let searchResults = null;
```
Comment (abp437) [inline src/toDoList.js]: Why is this set to null, initially?

File: src/toDoList.js [hunk 14]
```diff
@@ -2,39 +2,153 @@ import React from 'react';
...
+        searchResults = this.state.items.filter((value, i) => value.includes(e.target.value));
+        return this.setState(
```
Comment (abp437) [inline src/toDoList.js]: Why is this returned?

File: src/toDoList.js [hunk 15]
```diff
@@ -2,39 +2,153 @@ import React from 'react';
...
       <div>
         <div id="addTask">
-          <input className="input" size="40" type="text" placeholder="Enter Task" onChange={e => this.setState({ value: e.target.value })} value={this.state.value}/>
-          <button onClick={this.addItem} type="button" className="btn btn-primary">Add</button>
+          <input
+            className="input"
+            size="40"
+            type="text"
+            placeholder="Enter Task"
+            onChange={this.onChangeFunctions()}
```
Comment (abp437) [inline src/toDoList.js]: Why are the calls to `onChangeFunctions` ambiguous? If you choose to not provide arguments sometimes then provide default params in function definition.

File: src/toDoList.js [hunk 16]
```diff
@@ -2,39 +2,153 @@ import React from 'react';
...
     });
   }
 
-  displayList(){
-    return (this.state.items.slice(0).reverse().map((item,i) => <li key={i}>{item}</li>));
+  removeItem(id) {
+    let isCheck = false;
+    let newItems = this.state.items.filter((item, i) => {
+      if(item === this.state.searchResults[id] && !isCheck) {
+        isCheck = true;
```
Comment (abp437) [inline src/toDoList.js:64]: If you are using filter then, it should return something, and here it isn't.

**Review by abp437 — CHANGES_REQUESTED** [review]

File: src/toDoList.js (hunk 4 above)
Comment (parthshah51999) [reply src/toDoList.js]: removed

File: src/toDoList.js (hunk 3 above)
Comment (parthshah51999) [reply src/toDoList.js]: done

File: src/toDoList.js (hunk 2 above)
Comment (parthshah51999) [reply src/toDoList.js]: done - formatted code in toDoItem.js after refactoring

File: src/toDoList.js (hunk 9 above)
Comment (parthshah51999) [reply src/toDoList.js]: resolved

File: src/toDoList.js (hunk 11 above)
Comment (parthshah51999) [reply src/toDoList.js]: moved to separate search Bar component

File: src/toDoList.js (hunk 13 above)
Comment (parthshah51999) [reply src/toDoList.js]: corrected

File: src/toDoList.js (hunk 14 above)
Comment (parthshah51999) [reply src/toDoList.js]: corrected

File: src/toDoList.js (hunk 15 above)
Comment (parthshah51999) [reply src/toDoList.js]: separated out functions because search bar is created as a different component.

File: src/toDoList.js (hunk 12 above)
Comment (parthshah51999) [reply src/toDoList.js]: used indexOf instead

File: src/toDoList.js (hunk 16 above)
Comment (parthshah51999) [reply src/toDoList.js:64]: wanted to return  a single result from .filter, hence used isCheck flag in combination. Alternative way to do this ?

Comment (abp437) [reply src/toDoList.js:64]: Use For Loop.

File: src/toDoList.js [hunk 17]
```diff
@@ -127,4 +128,16 @@ class ToDoList extends React.Component {
   }
 }
 
+ToDoItem.propTypes = {
```
Comment (visha12l) [inline src/toDoList.js]: todoitem proptypes must be defined in Todoitem.js

File: src/toDoList.js [hunk 18]
```diff
@@ -127,4 +128,16 @@ class ToDoList extends React.Component {
...
+  searchResults: PropTypes.array,
+  removeItem: PropTypes.func.isRequired,
+  editItem: PropTypes.func.isRequired
+};
+
+SearchBar.defaultProps = {
```
Comment (visha12l) [inline src/toDoList.js]: define default props in Searchbar.js
